### PR TITLE
Add support for --{enable,disable}-{static,dynamic}

### PR DESCRIPTION
**Pull request by zyga**
Those configuration script options are now accepted and stored internally but
are not acted upon by the build system. Individually expanded Library.A,
Library.So and Library.DyLib templates become no-op when the corresponding
feature is disabled.

Fixes: https://github.com/zyga/zmk/issues/71
Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: tests/Library.A/Test.mk [hunk 1]
```diff
@@ -13,6 +14,10 @@ $(eval $(ZMK.isolateHostToolchain))
 %-silent-rules.log: ZMK.makeOverrides += Silent.Active=yes
 # Some logs have DESTDIR set to /destdir
 %-destdir.log: ZMK.makeOverrides += DESTDIR=/destdir
+# Some logs behave as if configure --enable-static was used
+%-enable-static-libs.log: ZMK.makeOverrides += Configure.StaticLibraries=yes
+# Some logs behave as if configure --disable -static was used
```
Comment (robertjdrab) [inline tests/Library.A/Test.mk:19]: NIT: s/--disable -static/--disable-static

File: tests/Library.A/Test.mk [hunk 2]
```diff
@@ -93,3 +98,10 @@ clean-destdir: clean-destdir.log
 	GREP -qFx 'rm -f ./libfoo.a-foo.o' <$<
 	# Cleaning removes the dependency files
 	GREP -qFx 'rm -f ./libfoo.a-foo.d' <$<
+
+all-enable-static-libs: all-enable-static-libs.log
+    # Configuring --enable-static enables compliation of static libraries.
```
Comment (robertjdrab) [inline tests/Library.A/Test.mk:103]: NIT: s/compliation/compilation/

File: tests/Library.A/Test.mk [hunk 3]
```diff
@@ -93,3 +98,10 @@ clean-destdir: clean-destdir.log
...
+	GREP -qFx 'ar -cr libfoo.a libfoo.a-foo.o' <$<
+all-disable-static-libs: all-disable-static-libs.log
+    # Configuring --disable-static disables compliation of static libraries.
```
Comment (robertjdrab) [inline tests/Library.A/Test.mk:106]: NIT: same as above

**Comment by zyga** [comment]
Thanks, I'll de-conflict and fix those shortly.

File: tests/Library.DyLib/Test.mk [hunk 4]
```diff
@@ -3,13 +3,18 @@
...
-    all-destdir install-destdir uninstall-destdir clean-destdir
+    all-destdir install-destdir uninstall-destdir clean-destdir \
+    all-enable-dynamic-libs all-disable-dynamic-libs
 
 $(eval $(ZMK.isolateHostToolchain))
 # Test logs will contain debugging messages
 %.log: ZMK.makeOverrides += DEBUG=library.dylib
 # Some logs have DESTDIR set to /destdir
 %-destdir.log: ZMK.makeOverrides += DESTDIR=/destdir
+# Some logs behave as if configure --enable-static was used
```
Comment (robertjdrab) [inline tests/Library.DyLib/Test.mk:14]: s/--enable-static/--enable-dynamic/

File: tests/Library.DyLib/Test.mk [hunk 5]
```diff
@@ -3,13 +3,18 @@
...
+%-enable-dynamic-libs.log: ZMK.makeOverrides += Configure.DynamicLibraries=yes
+# Some logs behave as if configure --disable -static was used
```
Comment (robertjdrab) [inline tests/Library.DyLib/Test.mk:16]: s/--disable -static/--disable-dynamic/

File: tests/Library.DyLib/Test.mk [hunk 6]
```diff
@@ -88,3 +93,10 @@ clean-destdir: clean-destdir.log
 	GREP -qFx 'rm -f ./libfoo.1.dylib-foo.d' <$<
 	GREP -qFx 'rm -f ./libbar.dylib-bar.o' <$<
 	GREP -qFx 'rm -f ./libbar.dylib-bar.d' <$<
+
+all-enable-dynamic-libs: all-enable-dynamic-libs.log
+    # Configuring --enable-dynamic enables compliation of dynamic libraries.
```
Comment (robertjdrab) [inline tests/Library.DyLib/Test.mk:98]: NIT:s/compliation/compilation/

File: tests/Library.DyLib/Test.mk [hunk 7]
```diff
@@ -88,3 +93,10 @@ clean-destdir: clean-destdir.log
...
+	GREP -qFx 'cc -dynamiclib -compatibility_version 1.0 -current_version 1.0 -o libfoo.1.dylib libfoo.1.dylib-foo.o' <$<
+all-disable-dynamic-libs: all-disable-dynamic-libs.log
+    # Configuring --disable-dynamic disables compliation of dynamic libraries.
```
Comment (robertjdrab) [inline tests/Library.DyLib/Test.mk:101]: NIT: same as above

File: tests/Library.So/Test.mk [hunk 8]
```diff
@@ -3,13 +3,18 @@
...
-    all-destdir install-destdir uninstall-destdir clean-destdir
+    all-destdir install-destdir uninstall-destdir clean-destdir \
+    all-enable-dynamic-libs all-disable-dynamic-libs
 
 $(eval $(ZMK.isolateHostToolchain))
 # Test logs will contain debugging messages
 %.log: ZMK.makeOverrides += DEBUG=library.so
 # Some logs have DESTDIR set to /destdir
 %-destdir.log: ZMK.makeOverrides += DESTDIR=/destdir
+# Some logs behave as if configure --enable-static was used
```
Comment (robertjdrab) [inline tests/Library.So/Test.mk:14]: s/-static/-dynamic/

File: tests/Library.So/Test.mk [hunk 9]
```diff
@@ -3,13 +3,18 @@
...
+%-enable-dynamic-libs.log: ZMK.makeOverrides += Configure.DynamicLibraries=yes
+# Some logs behave as if configure --disable -static was used
```
Comment (robertjdrab) [inline tests/Library.So/Test.mk:16]: s/ -static/-dynamic/

File: tests/Library.So/Test.mk [hunk 10]
```diff
@@ -88,3 +93,10 @@ clean-destdir: clean-destdir.log
 	GREP -qFx 'rm -f ./libfoo.so.1-foo.d' <$<
 	GREP -qFx 'rm -f ./libbar.so-bar.o' <$<
 	GREP -qFx 'rm -f ./libbar.so-bar.d' <$<
+
+all-enable-dynamic-libs: all-enable-dynamic-libs.log
+    # Configuring --enable-dynamic enables compliation of dynamic libraries.
```
Comment (robertjdrab) [inline tests/Library.So/Test.mk:98]: NIT:s/compliation/compilation/

File: tests/Library.So/Test.mk [hunk 11]
```diff
@@ -88,3 +93,10 @@ clean-destdir: clean-destdir.log
...
+	GREP -qFx 'cc -shared -Wl,-soname=libfoo.so.1 -o libfoo.so.1 libfoo.so.1-foo.o' <$<
+all-disable-dynamic-libs: all-disable-dynamic-libs.log
+    # Configuring --disable-dynamic disables compliation of dynamic libraries.
```
Comment (robertjdrab) [inline tests/Library.So/Test.mk:101]: NIT: same as above

**Review by robertjdrab — CHANGES_REQUESTED** [review]
Some comments need corrections.

File: tests/Library.So/Test.mk [hunk 12]
```diff
@@ -95,8 +95,8 @@ clean-destdir: clean-destdir.log
 	GREP -qFx 'rm -f ./libbar.so-bar.d' <$<
 
 all-enable-dynamic-libs: all-enable-dynamic-libs.log
-    # Configuring --enable-dynamic enables compliation of dynamic libraries.
+	# Configuring --enable-dynamic enables compliation of dynamic libraries.
```
Comment (robertjdrab) [inline tests/Library.So/Test.mk:98]: NIT: Two more s/compliation/compilation/
grep for 'compliation'

Comment (zyga) [reply tests/Library.So/Test.mk:98]: Uh, my vim settings are wrong and I had spellchecker off. Thanks for spotting that

**Review by robertjdrab — APPROVED** [review]
All good!

**Pull request merged**